### PR TITLE
support retry when downloading images

### PR DESCRIPTION
**Pull request by chloeyin**
**Purpose of the PR**

- support retry when downloading images #587  @northtyphoon  PTAL

### CODE REVIEW

File: graph/step.go [hunk 1]
```diff
@@ -52,27 +52,29 @@ func (c chanBool) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	Mounts           []*volume.Mount `yaml:"volumeMounts"`
+	Push             []string        `yaml:"push"`
+	Envs             []string        `yaml:"env"`
+	Expose           []string        `yaml:"expose"`
+	Ports            []string        `yaml:"ports"`
+	When             []string        `yaml:"when"`
+	ExitedWith       []int           `yaml:"exitedWith"`
+	ExitedWithout    []int           `yaml:"exitedWithout"`
+	Timeout          int             `yaml:"timeout"`
+	// DownloadRetries specifies how mant times a download in a step will be retried
```
Comment (binbin-li) [inline graph/step.go:74]: nit: how many(t)

Comment (chloeyin) [reply graph/step.go]: Thanks! I have fixed it.

File: builder/builder.go [hunk 2]
```diff
@@ -394,7 +394,7 @@ func (b *Builder) pullImageBeforeRun(ctx context.Context, cmdArgs string) error
 	if b.debug {
 		log.Printf("pull image args: %v\n", args)
 	}
-	return b.procManager.Run(ctx, args, nil, os.Stdout, os.Stdout, "")
+	return b.procManager.RunWithRetries(ctx, args, nil, os.Stdout, os.Stdout, "", retries, nil, 0, "")
```
[the remark this answers is not in the document]
Comment (binbin-li) [reply builder/builder.go]: Gotcha, that makes sense!

File: graph/step.go [hunk 3]
```diff
@@ -52,27 +52,30 @@ func (c chanBool) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	Push             []string        `yaml:"push"`
+	Envs             []string        `yaml:"env"`
+	Expose           []string        `yaml:"expose"`
+	Ports            []string        `yaml:"ports"`
+	When             []string        `yaml:"when"`
+	ExitedWith       []int           `yaml:"exitedWith"`
+	ExitedWithout    []int           `yaml:"exitedWithout"`
+	Timeout          int             `yaml:"timeout"`
+	// DownloadRetries specifies how many times a download in a step will be retried
+	DownloadRetries             int `yaml:"downloadRetries"`
```
Comment (northtyphoon) [inline graph/step.go:75]: May I suggest rename the settings to `CmdDownloadRetries` and `CmdDownloadRetryDelayInSeconds` to be more clear?

**Comment by northtyphoon** [comment]
Overall LGTM. Thanks. Have a suggestion on setting names. By the way, did you have a chance to test the retry like using a non-existence image in a cmd step?

**Comment by chloeyin** [comment]
> Overall LGTM. Thanks. Have a suggestion on setting names. By the way, did you have a chance to test the retry like using a non-existence image in a cmd step?

Yes, I have tried a simple yaml file
```
network:
  skipCreation:true
steps:
  - id: runner
    cmd: hello-world-not-exist:latest
    retries: 5
    pull: true
    cmdDownloadRetries: 5
    cmdDownloadRetryDelay: 5
```
By executing `bin/acb exec -f bin/acb.yaml --debug`
It will produce output like
```
2022/06/08 01:44:02 Rendered template:
network:
  skipCreation:true
steps:
  - id: runner
    cmd: hello-world-not-exist:latest
    retries: 5
    pull: true
    cmdDownloadRetries: 5
    cmdDownloadRetryDelay: 5
2022/06/08 01:44:02 Creating Docker network: acb_default_network, driver: 'bridge'
2022/06/08 01:44:02 Successfully set up Docker network: acb_default_network
2022/06/08 01:44:02 Setting up Docker configuration...
2022/06/08 01:44:03 Successfully set up Docker configuration
2022/06/08 01:44:03 Executing step ID: runner. Timeout(sec): 600, Working directory: '', Network: 'acb_default_network'
2022/06/08 01:44:03 Step specified pull. Performing an explicit pull...
2022/06/08 01:44:03 pull image args: [docker run --rm --volume /var/run/docker.sock:/var/run/docker.sock docker pull hello-world-not-exist:latest]
2022/06/08 01:44:03 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:11 Container failed during run: , waiting 5 seconds before retrying...
2022/06/08 01:44:16 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:20 Container failed during run: , waiting 5 seconds before retrying...
2022/06/08 01:44:25 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:29 Container failed during run: , waiting 5 seconds before retrying...
2022/06/08 01:44:34 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:39 Container failed during run: , waiting 5 seconds before retrying...
2022/06/08 01:44:44 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:48 Container failed during run: , waiting 5 seconds before retrying...
2022/06/08 01:44:53 Launching container with name:
Error response from daemon: pull access denied for hello-world-not-exist, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2022/06/08 01:44:58 Container failed during run: . No retries remaining.
failed to run step ID: runner: exit status 1
```

File: graph/step.go (hunk 3 above)
Comment (chloeyin) [reply graph/step.go]: Ok, I have fixed it.

**Review by northtyphoon — APPROVED** [review]
LGTM

**Pull request merged**